### PR TITLE
test(ios-agent): put the attribution walk behind a reader seam

### LINKED ISSUE #690: The netfilter's process attribution walk is the last part still untested
Parent: #607

The system extension is the layer that actually stops traffic, and nothing tests it. That is our gap,
not the platform's — worth stating plainly, because "Swift is hard to test" is the assumption that keeps
it open.

## What is actually there

XCTest exists and this repo already uses it. `packages/ios-agent/xctest-runner/TreeRunner/TreeServerTest.swift`
is an `XCTestCase` — but not a test:

```swift
func testServeTree() throws {
    let server = try TreeServer(port: port)
    server.start()
    RunLoop.current.run()   // block forever; the host kills the process
}
```

It is an XCTest shell used to keep a process alive inside a simulator. `ios-netfilter/project.yml`
declares two targets, `application` and `system-extension`, and no test target. Nothing prevents adding
one.

## Why it matters now

#654 and #685 both changed `Extension/Provider.swift` — the empty-rule short circuit in `handleNewFlow`
and per-device drop counters in `Heartbeat`. Both shipped with **reading and an adversarial review as
their only gate**. The agent half of the same change has fifteen tests.

The asymmetry is not academic: `handleNewFlow` decides whether a connection is allowed, and a mistake
there is either a tester signing off on offline behaviour that never happened, or the user's own browser
being cut.

## What is testable, and what is not

Worth splitting, because the honest scope is narrower than "test the provider" and wider than "nothing":

- **`Heartbeat` is plain Swift with no NetworkExtension dependency** — the counters, `renderLocked`'s
  hand-built JSON, `dueLocked`'s rate arithmetic, `pulseSeconds(enforcing:)`, the path probing. This is
  where the last two changes landed, and all of it is unit-testable today.
- **`handleNewFlow` is not, easily** — it takes an `NEFilterFlow`, which is awkward to construct. Its
  logic could be extracted behind a seam that takes `(pid, asid, rule)` and returns a verdict, which
  would make the empty-rule skip and the attribution branches testable without NE.
- **`attribute(pid)`** walks real processes; it would need either a seam or a live target.

## The cost to decide

CI runs on `ubuntu-latest` for all three jobs, so Swift tests need a macOS runner or they are local-only
— present but not a gate, which is a shape this repo has been bitten by before. That is the decision
this issue exists for: add the target and accept a local-only suite, add a macOS runner, or decide the
provider is reviewed rather than tested and write *that* down where the next person changing it will
look.

## Where to start

`packages/ios-agent/ios-netfilter/project.yml` (add the target), `Extension/Provider.swift`
(`Heartbeat` first — it needs no seam), and `.github/workflows/ci.yml` for whichever way the runner
question goes.

### PR DESCRIPTION
**Pull request by jo-duchan**
## Summary

The last of #690's three bullets, and the issue closes with it.

`attribute(pid)` is the climb that decides whose traffic a flow is — a simulator's, the Mac's, or
nothing anyone can name. Nothing tested it, because it reaches the kernel three ways:
`sysctl(KERN_PROC)` for a process's parent, `proc_pidpath` for its executable, `KERN_PROCARGS2` for
its arguments. Those arrive as a `ProcessReader` now and the climb is `attributeWalk`.

**Closures rather than values, which is this program's most repeated lesson.** A read turned into an
argument is evaluated before the branch that needs it, and that has happened three times in these
files: `flowShape` onto every flow, `NSBundle.mainBundle` into every process the simulator starts,
`rl.rlim_cur` when `getrlimit` had failed. A closure is not evaluated until it is called.

Twelve cases hold what the climb decides: that the stop is `ppid <= 1` and not `== 1`, because a
process reparented to the kernel reports zero; that an unreadable executable path still reaches the
arguments, since refusing there reports a simulator's traffic as the Mac's and lets an offline device
keep talking; that a cache hit skips the argument read; that a failed read is `unresolved` and names
the pid it stopped at rather than being logged as a host flow the way it was before #642; and the
depth bound from both sides.

The review confirmed the refactor mechanically — it stripped comments from both function bodies,
renamed the moved symbols, and diffed them byte-identical — and then broke the coverage claim in three
places:

- **The identity the walk hands the cache had nothing holding it.** `UDIDCache` has a mutation for its
  dictionary's key; the caller's argument had none, so discarding the start time reintroduced the
  pid-reuse bug with all 79 tests green.
- **`pathReads` was a counter nothing read**, so calling `proc_pidpath` at every level of the climb
  passed.
- **"The bound cannot be mutated" was asserted, not measured.** With XCTest's execution-time allowance
  a hang becomes a reported timeout, the run continues, and the boundary case kills it by name. It
  costs 87s against ~13s for a normal mutation, and nothing on a green run.

80 tests, 82 mutations, all killed.

## Checklist

- [x] Tests written and passing
- [x] No `any`
- [x] Interface changes land in `agent-core` first — n/a
- [x] No sensitive info (tokens, paths, credentials)

## Related `.work/` docs

- `.work/2026-09-05-netfilter-pure-logic-tests-plan.md` — the program this closes, with what each
  round taught.
- `.work/reviews/test__netfilter-attribution-walk.md` — one channel, five findings, none deferred.

Closes #690.

## Summary by CodeRabbit

- **Refactor**
  - Improved iOS network-flow attribution across simulator, host, and unresolved process cases.
  - Added safeguards for deep or incomplete process relationships.

- **Bug Fixes**
  - Improved handling of failed process lookups and missing simulator identification data.
  - Ensured repeated flow checks reuse cached attribution when appropriate.

- **Tests**
  - Expanded coverage for attribution, caching, boundary conditions, and failure scenarios.

- **Chores**
  - Updated iOS agent build versions and CI test-run configuration.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -194,7 +194,7 @@ jobs:
   #
   # **`--mutate`, not a plain run.** Many of the tests assert that something is *not* allowed, and a
   # test asserting absence passes when nothing happens — that is its definition. The flag breaks the
-  # sources seventy ways and requires each one to fail a test, which is the only thing that makes
+  # sources seventy-eight ways and requires each one to fail a test, which is the only thing that makes
```
[the remark this answers is not in the document]
Comment (jo-duchan) [reply .github/workflows/ci.yml]: Correct, and the drift happened inside this PR: `253e94cf` added eight mutations and updated the comment to seventy-eight; the review-fix commit `0dc8c9f9` added four more and did not. `ios-agent/AGENTS.md` was staler still, at forty-one.

Fixed in `50e8a431` — eighty-two and eighty-three, in both files.

Two things beyond the numbers.

**A fourth sentence would not have helped.** That paragraph already warns about this exact drift: it names a version of itself that "said thirty-six when it was thirty-four", and `AGENTS.md` spells out the grep that gets it right (`grep -cE '^mutate "'`, not `grep -c '^mutate '`, which counts the function definition). Both were written by someone who had just been bitten, and it drifted twice more. That is `contributing/test-and-guard-coverage.md` rule 1 — writing the lesson down reads as having applied it — so this got `scripts/__tests__/mutationCountsStated.test.mjs` instead. Six mutations prove it bites: moving the count in either script, either stated number, the off-by-one between mutations and launches, a reword that breaks an anchor, and a second bare `mutate ` line that would corrupt the count the way the warning describes.

Writing it caught one thing worth mentioning: the first version tripped `checksWalkDisk`, because a regex containing the literal `sources (` reads as a module enumerating the tree. The anchors now match whitespace runs — which they needed anyway, since `ci.yml`'s prose wraps across a `# ` marker.

**The timing paragraph was extrapolated and is now measured.** The run of `0dc8c9f9` did eighty-three launches in 17m29s with the Release build 23s behind it — 18m03s for the job, 1.94x its 35-minute timeout, against the "about a minute" the comment claimed for that build. Launch cost is not one number either: 9.7–11.75s at sixteen launches, 13.5s at forty-two, 12.6s at eighty-three. The paragraph now says to extrapolate from the slowest of those rather than the luckiest.
